nebullvm/operations/optimizations/base.py

```diff
@@ -0,0 +1,242 @@
+import abc
+from tempfile import TemporaryDirectory
+from typing import List, Callable, Union, Tuple, Any
+
+from nebullvm.config import CONSTRAINED_METRIC_DROP_THS
+from nebullvm.operations.base import Operation
+from nebullvm.operations.measures.measures import MetricDropMeasure
+from nebullvm.operations.measures.utils import (
+    compute_relative_difference,
+    compute_optimized_running_time,
+)
+
+from nebullvm.optional_modules.tensorflow import tensorflow as tf
+from nebullvm.optional_modules.torch import torch
+from nebullvm.tools.base import (
+    ModelCompiler,
+    QuantizationType,
+    OptimizationTime,
+    ModelParams,
+    DeepLearningFramework,
+    ModelCompressor,
+)
+from nebullvm.tools.data import DataManager
+from nebullvm.tools.feedback_collector import FEEDBACK_COLLECTOR
+from nebullvm.tools.transformations import MultiStageTransformation
+
+
+class Optimizer(Operation, abc.ABC):
+    def __init__(self):
+        super().__init__()
+        self.optimized_models = []
+        self.source_dl_framework = None
+        self.pipeline_dl_framework = None
+        self.compiler_ops = {}
+        self.build_inference_learner_ops = {}
+        self.validity_check_op = MetricDropMeasure()
+
+    def execute(
+        self,
+        model: Any,
+        input_data: DataManager,
+        optimization_time: OptimizationTime,
+        metric_drop_ths: str,
+        metric: Callable,
+        model_params: ModelParams,
+        model_outputs: List[Tuple[Any, ...]],
+        ignore_compilers: List[ModelCompiler],
+        ignore_compressors: List[ModelCompressor],
+        source_dl_framework: DeepLearningFramework,
+    ):
+        self.source_dl_framework = source_dl_framework
+
+        # TODO: implement and select compressors from hardware
+
+        compilers = self._select_compilers_from_hardware()
+        (
+            self.compiler_ops,
+            self.build_inference_learner_ops,
+        ) = self._load_compilers(
+            ignore_compilers=ignore_compilers,
+            compilers=compilers,
+        )
+        self.optimize(
+            model=model,
+            input_data=input_data,
+            optimization_time=optimization_time,
+            metric_drop_ths=metric_drop_ths,
+            metric=metric,
+            model_params=model_params,
+            model_outputs=model_outputs,
+            ignore_compilers=ignore_compilers,
+        )
+
+    @abc.abstractmethod
+    def _select_compilers_from_hardware(self):
+        raise NotImplementedError()
+
+    def _load_compilers(
+        self,
+        ignore_compilers: List[ModelCompiler],
+        compilers: List[ModelCompiler],
+    ):
+        from nebullvm.operations.optimizations.optimizers import (
```

Comment:
why are we importing the optimizers inside the function?

Reply:
because there were circular import issues